### DISCUSSION TITLE
Charging a storage unit using a particular carrier

### DISCUSSION THREAD

**Discussion by indiatce** [pypsa-meets-earth/pypsa-earth#899] [Q&A]
I want to charge my storage unit by a particular carrier, which can be either grid or solar or wind. How to apply constraints in the CSV files for the storage units such as 'storage_units', 'storage_units-inflow' and 'storage_units-state_of_charge_set', or in the code. I tried to connect the storage unit with uni-directional link, but the links are not working in unidirectional mode. Can you please suggest me, how and where can I implement such constraints so that my storage unit gets charged by a particular carrier, even if it is interconnected with the other buses (both generator and load).

**Comment by davide-f** [comment]
Hello @indiatce :D and welcome!

So, to do what you desire to do, I'd recommend you to add a constrain with extra_functionalities to make sure that the charging power of your converter (be it the storage_unit or the charging link) is lower than the available production by the carrier of interest in every time step.
Examples on how to use extra_functionalities with linopy are discussed [here](https://pypsa.readthedocs.io/en/latest/examples/optimization-with-linopy-migrate-extra-functionalities.html#Optimization-with-Linopy---Migrate-Extra-Functionalities)

To create unidirectional links, you need to specify p_max/min_pu accordingly: one being 0 and the other one +1 or -1, depending on what bus0 and 1 you specified.
If you specify for both links the same bus0 and bus1, then, one link must have p_max_pu=1 (default value) and p_min_pu=0 (default value) and the other link: p_max_pu=0, p_min_pu=-1.
However, to achieve the goal that you aim to, the only way I see is with extra_functionalities

**Reply by indiatce** [reply]
Thank you so much for the reply. I will go through the ''extra functionalities" and will try
